src/settings.ts

```diff
@@ -0,0 +1,104 @@
+  (async () => migrateSettings(settingsFromConfigFiles, secureStorage))();
+  return settingsFromConfigFiles;
+}
+
+async function migrateSettings(settings: VSCode.WorkspaceConfiguration, secureStorage?: SonarLintSecretStorage) {
+  if (secureStorage !== undefined) {
+    const sqConnections = settings.get<SqConnection[]>('connectedMode.connections.sonarqube');
+    if (await hasUnmigratedConnections(sqConnections, secureStorage)) {
+      suggestMigrationToSecureStorage(sqConnections, secureStorage);
+    }
+  }
+}
+
+async function suggestMigrationToSecureStorage(sqConnections: SqConnection[], secureStorage: SonarLintSecretStorage) {
+  const remindMeLaterAction = 'Ask me later';
+  const migrateToSecureStorageAction = 'Migrate connection settings to secure storage';
+  const message = `SonarLint found SonarQube token in settings file. Do you want to migrate them to secure storage?`;
+  const selection = await VSCode.window.showWarningMessage(message, migrateToSecureStorageAction, remindMeLaterAction);
+  switch (selection) {
+    case migrateToSecureStorageAction:
+      await addTokensFromSettingsToSecureStorage(sqConnections, secureStorage);
+      break;
+    case remindMeLaterAction:
+  }
+}
+
+function getSettingsFileConfig(): VSCode.WorkspaceConfiguration {
+  return VSCode.workspace.getConfiguration('sonarlint');
+}
+
+async function addTokensFromSettingsToSecureStorage(sqConnections: SqConnection[],
+                                                    secureStorage: SonarLintSecretStorage) {
+  await Promise.all(
+    sqConnections.map(async c => {
+      if (c.token !== undefined && !await secureStorage.hasTokenForServer(c.serverUrl)) {
+        await secureStorage.storeServerToken(c.serverUrl, c.token);
+        c.token = undefined;
+      }
+    })
+  );
+  return VSCode.workspace.getConfiguration()
+    .update('sonarlint.connectedMode.connections.sonarqube', sqConnections, VSCode.ConfigurationTarget.Global);
+}
+
+export class SonarLintSecretStorage {
+
+  private static instance: SonarLintSecretStorage;
+
+  constructor(private readonly secretStorage: VSCode.SecretStorage) {
+  }
+
+  static init(context: ExtensionContext): void {
+    SonarLintSecretStorage.instance = new SonarLintSecretStorage(context.secrets);
+  }
+
+  static get getInstance(): SonarLintSecretStorage {
+    return SonarLintSecretStorage.instance;
+  }
+
+  /**
+   *
+   * @param serverId SonarQube URL or SonarCloud organization ID
+   * @param token auth token
+   */
+  async storeServerToken(serverId: string, token: string): Promise<void> {
+    if (token) {
+      this.secretStorage.store(serverId, token);
+    }
+  }
+
+  async getServerToken(serverId: string): Promise<string | undefined> {
+    return this.secretStorage.get(serverId);
+  }
+
+  async hasTokenForServer(serverId: string): Promise<boolean> {
+    return (await this.getServerToken(serverId)) !== undefined;
+  }
+
+}
+
+type SqConnection = { serverUrl: string, token: string };
```

Comment:
It can also have `connectionId? : string`

Reply:
Changed to { serverId?: string, serverUrl?: string, organizationKey?: string, token?: string }